src/entities/nestedExit/getNestedExitCalls.ts

```diff
@@ -0,0 +1,193 @@
+import { Token } from '../token';
+import { NestedExitInput, NestedExitCall } from './types';
+import { NestedPool, NestedPoolState } from '../types';
+import { TokenAmount } from '../tokenAmount';
+import { Address } from '../../types';
+
+export const getNestedExitCalls = (
+    {
+        bptAmountIn,
+        chainId,
+        accountAddress,
+        tokenOut,
+        useNativeAssetAsWrappedAmountOut = false,
+        toInternalBalance = false,
+    }: NestedExitInput,
+    { pools }: NestedPoolState,
+): { bptAmountIn: TokenAmount; calls: NestedExitCall[] } => {
+    const isProportional = tokenOut === undefined;
+    let poolsTopDown: NestedPool[];
+    let calls: NestedExitCall[];
+
+    if (isProportional) {
+        /**
+         * Overall logic to build sequence of proportional exit calls:
+         * 1. Go from top pool to bottom filling out input amounts and output refs
+         * 2. Inputs will be bptAmountIn provided or output of the previous level
+         * 3. Output at bottom level is the amountsOut
+         */
+
+        // sort pools by descending level
+        poolsTopDown = pools.sort((a, b) => b.level - a.level);
+
+        calls = getProportionalExitCalls(
+            poolsTopDown,
+            chainId,
+            useNativeAssetAsWrappedAmountOut,
+            accountAddress,
+            bptAmountIn,
+            toInternalBalance,
+        );
+    } else {
+        /**
+         * Overall logic to build sequence of single token exit calls:
+         * 1. Go BOTTOM-UP building exit path to tokenOut
+         * 2. Go through exit path filling out input amounts and output refs
+         * 3. Inputs will be bptAmountIn provided or output of the previous level
+         * 4. Output at bottom level is the amountOut
+         */
+
+        // sort pools by descending level
+        poolsTopDown = pools.sort((a, b) => b.level - a.level);
+        const topPool = poolsTopDown[0];
+
+        // Go BOTTOM-UP building exit path to tokenOut
+        const exitPath: NestedPool[] = [];
+        let tokenOutByLevel = tokenOut;
+        while (tokenOutByLevel !== topPool.address) {
+            const currentPool = poolsTopDown.find(
```

Comment:
Just so we cover it - this is always finding the first pool. So in the case there are more than 1 pools with the token it should still work but it'll always find the same path? Can you see any issues here?

Reply:
The code will still work and I'd say that it should find the same path if the api response returns pools in the same order (which I'm not sure is enforced).
But taking a step back, do we want to allow this kind of behavior?
I'd say that the api should not provide nested join/exit data for pools that are outside of the scope.
In other words, I'd say it's the api responsibility to prevent cases like that from ever reaching the SDK.
What do you think?

Reply:
I'm not 100% on this. I kind of feel like we should have a check/validation early in the code and throw if we are hitting any of the constraints. I'll make a note to discuss this during the stand up to get Daniels view.